I ran the tests I added to https://github.com/tenstorrent/tt-metal/pull/18104 up to now on a WH N150 (x1) and got:
```
PASSED tests/ttnn/unit_tests/operations/test_matmul.py::test_padded_2d_matmul[tile_count=1376-side=height]
PASSED tests/ttnn/unit_tests/operations/test_matmul.py::test_padded_2d_matmul[tile_count=1376-side=width]
PASSED tests/ttnn/unit_tests/operations/test_matmul.py::test_padded_2d_matmul[tile_count=1375-side=height]
PASSED tests/ttnn/unit_tests/operations/test_matmul.py::test_padded_2d_matmul[tile_count=1375-side=width]

FAILED tests/ttnn/unit_tests/operations/test_matmul.py::test_padded_1d_matmul[has_program_config=True] - AssertionError: 0.952219109365194
FAILED tests/ttnn/unit_tests/operations/test_matmul.py::test_padded_1d_matmul[has_program_config=False] - AssertionError: 0.952219109365194

FAILED tests/ttnn/unit_tests/operations/test_matmul.py::test_block_sharded[transpose_mcast=True-k_size=768-m_size=1920-n_size=768] - AssertionError: 0.08211901437793893
FAILED tests/ttnn/unit_tests/operations/test_matmul.py::test_block_sharded[transpose_mcast=False-k_size=768-m_size=1920-n_size=768] - AssertionError: 0.08397262611314017
```

--- a/tests/ttnn/unit_tests/operations/test_matmul.py
+++ b/tests/ttnn/unit_tests/operations/test_matmul.py
@@ -1185,6 +1185,168 @@ def test_matmul_1d_multiple_output_blocks_per_core(
     assert device.num_program_cache_entries() == 1
 
 
+@pytest.mark.parametrize("side", ["height", "width"])
+@pytest.mark.parametrize("tile_count", [1376, 1375])
+def test_padded_2d_matmul(device, side, tile_count):
+    """
+    This test checks that when the program config specifies per_core_M and per_core_N
+    which would multiply out to be larger than the true shape of the output, matmul
+    does not clobber memory outside the shape of the output.
+    """
+    compute_grid_size = device.compute_with_storage_grid_size()
+    grid_size = [compute_grid_size.x, compute_grid_size.y]
+    if grid_size[1] < 8:
+        pytest.skip("device does not have 8x8 grid")
+
+    if side == "height":
+        M = tile_count * 32
+        K = 256
+        N = 32
+        out_block_h = 11
+        out_block_w = 1
+        per_core_M = 176
+        per_core_N = 1
+    else:
+        M = 32
+        K = 256
+        N = tile_count * 32
+        out_block_h = 1
+        out_block_w = 11
+        per_core_M = 1
+        per_core_N = 176
+    torch.manual_seed(0)
+    program_config = ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
+        compute_with_storage_grid_size=(8, 8),
+        in0_block_w=1,
+        out_block_h=out_block_h,
+        out_block_w=out_block_w,
+        out_subblock_h=1,
+        out_subblock_w=1,
+        per_core_M=per_core_M,
+        per_core_N=per_core_N,
+        transpose_mcast=False,
+        fused_activation=None,
+        fuse_batch=False,
+    )
+
+    torch_act = torch.randn([1, 1, M, K], dtype=torch.bfloat16)
+    torch_weight = torch.randn([1, 1, K, N], dtype=torch.bfloat16)
+    # Allocate tensors above and below where the output will be
+    X = 2**8
+    dummy_lower = torch.full([1, 1, X, X], 2)
+    dummy_out = torch.zeros([1, 1, M, N])
+    dummy_upper = torch.full([1, 1, X, X], 4)
+
+    act = ttnn.from_torch(torch_act, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    weight = ttnn.from_torch(torch_weight, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    lower_tt = ttnn.from_torch(dummy_lower, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    out_tt = ttnn.from_torch(dummy_out, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    upper_tt = ttnn.from_torch(dummy_upper, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    # Free up dummy output tensor so matmul will allocate output there
+    ttnn.deallocate(out_tt)
+    output_tensor = ttnn.matmul(
+        act,
+        weight,
+        program_config=program_config,
+        compute_kernel_config=ttnn.WormholeComputeKernelConfig(
+            math_fidelity=ttnn.MathFidelity.HiFi2, math_approx_mode=False, fp32_dest_acc_en=True, packer_l1_acc=False
+        ),
+    )
+    lower = ttnn.to_torch(lower_tt).float()
+    upper = ttnn.to_torch(upper_tt).float()
+    # Check that the tensors above and below the output are unchanged
+    torch_output_tensor = torch.matmul(torch_act, torch_weight)
+    output_tensor = ttnn.to_torch(output_tensor)
+    pcc = 0.999
+    assert_with_pcc(torch_output_tensor, output_tensor, pcc)
+    assert torch.all(lower == 2)
+    assert torch.all(upper == 4)
+
+
+@pytest.mark.parametrize("side", ["height", "width"])
+@pytest.mark.parametrize(
+    "has_program_config",
+    [
+        True,
+    ],
+)
+def test_padded_1d_matmul(device, side, has_program_config):
+    # pytest.skip("#17491: last tile bounds are not set up properly")
+    if side == "height":
+        M = 10369
+        K = 96
+        N = 1152
+        out_block_h = 21
+        out_block_w = 9
+        out_subblock_h = 3
+        out_subblock_w = 1
+        per_core_M = 21
+        per_core_N = 36
+        mcast_in0 = False
+    else:
+        M = 1152
+        K = 96
+        N = 10369
+        out_block_h = 9
+        out_block_w = 21
+        out_subblock_h = 1
+        out_subblock_w = 3
+        per_core_M = 36
+        per_core_N = 21
+        mcast_in0 = True
+    if has_program_config:
+        program_config = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
+            compute_with_storage_grid_size=(4, 4),
+            in0_block_w=1,
+            out_block_h=out_block_h,
+            out_block_w=out_block_w,
+            out_subblock_h=out_subblock_h,
+            out_subblock_w=out_subblock_w,
+            per_core_M=per_core_M,
+            per_core_N=per_core_N,
+            mcast_in0=mcast_in0,
+            fused_activation=None,
+            fuse_batch=True,
+        )
+    else:
+        program_config = None
+
+    torch.manual_seed(0)
+    pcc = 0.999
+    torch_act = torch.randn([1, 1, M, K], dtype=torch.float16)
+    torch_weight = torch.randn([1, 1, K, N], dtype=torch.float16)
+    # Allocate tensors above and below where the output will be
+    X = 2**8
+    dummy_lower = torch.full([1, 1, X, X], 2)
+    dummy_out = torch.zeros([1, 1, M, N])
+    dummy_upper = torch.full([1, 1, X, X], 4)
+
+    act = ttnn.from_torch(torch_act, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    weight = ttnn.from_torch(torch_weight, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    lower_tt = ttnn.from_torch(dummy_lower, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    out_tt = ttnn.from_torch(dummy_out, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    upper_tt = ttnn.from_torch(dummy_upper, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.bfloat16)
+    # Free up dummy output tensor so linear will allocate output there
+    ttnn.deallocate(out_tt)
+    output_tensor = ttnn.matmul(
+        act,
+        weight,
+        program_config=program_config,
+        compute_kernel_config=ttnn.WormholeComputeKernelConfig(
+            math_fidelity=ttnn.MathFidelity.HiFi2, math_approx_mode=False, fp32_dest_acc_en=True, packer_l1_acc=False
+        ),
+    )
+    lower = ttnn.to_torch(lower_tt).float()
+    upper = ttnn.to_torch(upper_tt).float()
+    # Check that the tensors above and below the output are unchanged
+    torch_output_tensor = torch.matmul(torch_act, torch_weight)
+    output_tensor = ttnn.to_torch(output_tensor)
+    # assert_with_pcc(torch_output_tensor, output_tensor, pcc)
+    assert torch.all(lower == 2)
+    assert torch.all(upper == 4)
+    # assert_with_pcc(torch_output_tensor, output_tensor, pcc)
+
+
 # fmt: off
 @pytest.mark.skipif(is_wormhole_b0() or is_blackhole(), reason="Unsupported on WH and BH")
 @pytest.mark.parametrize("m_size,k_size,n_size", [

--- a/ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_receiver_writer_padding.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_receiver_writer_padding.cpp
@@ -20,6 +20,8 @@ void kernel_main() {
     uint32_t out_tensor_start_tile_id = get_arg_val<uint32_t>(rt_args_idx++);
 
     // padding args (WRITER)
+    const uint32_t last_num_blocks_h_dim = get_arg_val<uint32_t>(rt_args_idx++);
+    const uint32_t last_num_blocks_w_dim = get_arg_val<uint32_t>(rt_args_idx++);
     const uint32_t out_num_nonzero_subblocks_h = get_arg_val<uint32_t>(rt_args_idx++);
     const uint32_t out_last_num_nonzero_subblocks_h = get_arg_val<uint32_t>(rt_args_idx++);
     const uint32_t out_last_subblock_h = get_arg_val<uint32_t>(rt_args_idx++);
@@ -140,12 +142,14 @@ void kernel_main() {
 
 #ifndef OUT_SHARDED
                 // WRITER
+                uint32_t num_blocks_h_dim_ = bh >= last_num_blocks_h_dim - 1 ? last_num_blocks_h_dim : num_blocks_h_dim;
+                uint32_t num_blocks_w_dim_ = bw >= last_num_blocks_w_dim - 1 ? last_num_blocks_w_dim : num_blocks_w_dim;
                 uint32_t out_num_nonzero_subblocks_h_ = out_num_nonzero_subblocks_h;
                 uint32_t out_num_nonzero_subblocks_w_ = out_num_nonzero_subblocks_w;
-                if (bh == num_blocks_h_dim - 1) {
+                if (bh == num_blocks_h_dim_ - 1) {
                     out_num_nonzero_subblocks_h_ = out_last_num_nonzero_subblocks_h;
                 }
-                if (bw == num_blocks_w_dim - 1) {
+                if (bw == num_blocks_w_dim_ - 1) {
                     out_num_nonzero_subblocks_w_ = out_last_num_nonzero_subblocks_w;
                 }
                 uint32_t out_tensor_sbh_start_tile_id = out_tensor_current_w_dim_block_tile_id;
@@ -157,10 +161,10 @@ void kernel_main() {
                         uint32_t out_subblock_h_ = out_subblock_h;
                         uint32_t out_subblock_w_ = out_subblock_w;
                         uint32_t subblock_tiles_addr_skip = 0;
-                        if (bh == num_blocks_h_dim - 1 && sbh == out_num_nonzero_subblocks_h - 1) {
+                        if (bh == num_blocks_h_dim_ - 1 && sbh == out_num_nonzero_subblocks_h_ - 1) {
                             out_subblock_h_ = out_last_subblock_h;
                         }
-                        if (bw == num_blocks_w_dim - 1 && sbw == out_num_nonzero_subblocks_w - 1) {
+                        if (bw == num_blocks_w_dim_ - 1 && sbw == out_num_nonzero_subblocks_w_ - 1) {
                             out_subblock_w_ = out_last_subblock_w;
                             subblock_tiles_addr_skip = padded_subblock_tiles_addr_skip;
                         }
@@ -171,7 +175,9 @@ void kernel_main() {
                         for (uint32_t h = 0; h < out_subblock_h_; ++h) {
                             uint32_t out_tensor_tile_id = out_tensor_sb_row_start_tile_id;
                             for (uint32_t w = 0; w < out_subblock_w_; ++w) {
-                                noc_async_write_tile(out_tensor_tile_id, s, l1_read_addr);
+                                if (bh < num_blocks_h_dim_ && bw < num_blocks_w_dim_) {
+                                    noc_async_write_tile(out_tensor_tile_id, s, l1_read_addr);
+                                }
 
                                 l1_read_addr += output_single_tile_size_bytes;
 
@@ -188,14 +194,14 @@ void kernel_main() {
                         out_tensor_sbw_start_tile_id += out_tensor_next_subblock_stride_w;
                     }
                     // Pop fully padded subblocks along the row
-                    if (bw == num_blocks_w_dim - 1) {
+                    if (bw == num_blocks_w_dim_ - 1) {
                         cb_wait_front(cb_id_out0, padded_block_tiles_w_skip);
                         cb_pop_front(cb_id_out0, padded_block_tiles_w_skip);
                     }
                     out_tensor_sbh_start_tile_id += out_tensor_next_subblock_stride_h;
                 }
                 // Pop row(s) of fully padded subblocks
-                if (bh == num_blocks_h_dim - 1) {
+                if (bh == num_blocks_h_dim_ - 1) {
                     cb_wait_front(cb_id_out0, padded_block_tiles_h_skip);
                     cb_pop_front(cb_id_out0, padded_block_tiles_h_skip);
                 }

--- a/ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_sender_writer_padding.cpp
@@ -28,6 +28,7 @@ void kernel_main() {
     // padding args (READER)
     const uint32_t last_block_w = get_arg_val<uint32_t>(rt_args_idx++);
     // padding args (WRITER)
+    const uint32_t last_num_blocks_w_dim = get_arg_val<uint32_t>(rt_args_idx++);
     const uint32_t out_num_nonzero_subblocks_h = get_arg_val<uint32_t>(rt_args_idx++);
     const uint32_t out_last_subblock_h = get_arg_val<uint32_t>(rt_args_idx++);
     const uint32_t padded_block_tiles_h_skip = get_arg_val<uint32_t>(rt_args_idx++);
@@ -420,9 +421,10 @@ void kernel_main() {
 
 #ifndef OUT_SHARDED
                 // WRITER
+                uint32_t num_blocks_w_dim_ = bw >= last_num_blocks_w_dim - 1 ? last_num_blocks_w_dim : num_blocks_w_dim;
                 uint32_t out_num_nonzero_subblocks_h_ = out_num_nonzero_subblocks_h;
                 uint32_t out_num_nonzero_subblocks_w_ = out_num_nonzero_subblocks_w;
-                if (bw == num_blocks_w_dim - 1) {
+                if (bw == num_blocks_w_dim_ - 1) {
                     out_num_nonzero_subblocks_w_ = out_last_num_nonzero_subblocks_w;
                 }
                 uint32_t out_tensor_sbh_start_tile_id = out_tensor_current_w_dim_block_tile_id;
@@ -437,7 +439,7 @@ void kernel_main() {
                         if (bh == num_blocks_h_dim - 1 && sbh == out_num_nonzero_subblocks_h - 1) {
                             out_subblock_h_ = out_last_subblock_h;
                         }
-                        if (bw == num_blocks_w_dim - 1 && sbw == out_num_nonzero_subblocks_w - 1) {
+                        if (bw == num_blocks_w_dim_ - 1 && sbw == out_num_nonzero_subblocks_w - 1) {
                             out_subblock_w_ = out_last_subblock_w;
                             subblock_tiles_addr_skip = padded_subblock_tiles_addr_skip;
                         }
@@ -448,7 +450,9 @@ void kernel_main() {
                         for (uint32_t h = 0; h < out_subblock_h_; ++h) {
                             uint32_t out_tensor_tile_id = out_tensor_sb_row_start_tile_id;
                             for (uint32_t w = 0; w < out_subblock_w_; ++w) {
-                                noc_async_write_tile(out_tensor_tile_id, s, l1_read_addr);
+                                if (bw < num_blocks_w_dim_) {
+                                    noc_async_write_tile(out_tensor_tile_id, s, l1_read_addr);
+                                }
 
                                 l1_read_addr += output_single_tile_size_bytes;
 
@@ -464,7 +468,7 @@ void kernel_main() {
                         out_tensor_sbw_start_tile_id += out_tensor_next_subblock_stride_w;
                     }
                     // Pop fully padded subblocks along the row
-                    if (bw == num_blocks_w_dim - 1) {
+                    if (bw == num_blocks_w_dim_ - 1) {
                         cb_wait_front(cb_id_out0, padded_block_tiles_w_skip);
                         cb_pop_front(cb_id_out0, padded_block_tiles_w_skip);
                     }

--- a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
@@ -718,23 +718,17 @@ operation::ProgramWithCallbacks create_program_mcast_in0(
             in3_CB_size);
     }
 
-    // Parameters for last row, col, or block
-    uint32_t last_per_core_M = M % per_core_M == 0 ? per_core_M : M % per_core_M;
+    // Parameters for last row, col, or block, no need to re-calc h-dim since there's no split on height
     uint32_t last_per_core_N = N % per_core_N == 0 ? per_core_N : N % per_core_N;
-    uint32_t last_out_block_h = last_per_core_M % out_block_h == 0 ? out_block_h : last_per_core_M % out_block_h;
     uint32_t last_out_block_w = last_per_core_N % out_block_w == 0 ? out_block_w : last_per_core_N % out_block_w;
-    uint32_t last_block_num_nonzero_subblocks_h = (last_out_block_h - 1) / out_subblock_h + 1;
+    uint32_t last_out_num_blocks_w = (last_per_core_N - 1) / out_block_w + 1;
     uint32_t last_block_num_nonzero_subblocks_w = (last_out_block_w - 1) / out_subblock_w + 1;
-    uint32_t last_subblock_of_last_block_h =
-        last_out_block_h % out_subblock_h == 0 ? out_subblock_h : last_out_block_h % out_subblock_h;
     uint32_t last_subblock_of_last_block_w =
         last_out_block_w % out_subblock_w == 0 ? out_subblock_w : last_out_block_w % out_subblock_w;
     uint32_t last_block_padded_subblock_tiles_addr_skip =
         output_single_tile_size * (out_subblock_w - last_subblock_of_last_block_w);
     uint32_t last_block_padded_block_tiles_w_skip =
         (out_subblock_w * out_subblock_h) * (out_block_w / out_subblock_w - last_block_num_nonzero_subblocks_w);
-    uint32_t last_block_padded_block_tiles_h_skip =
-        (out_block_h / out_subblock_h - last_block_num_nonzero_subblocks_h) * (out_block_w * out_subblock_h);
 
     CoreCoord start_core_noc = top_left_core_physical;
     CoreCoord end_core_noc = bottom_right_core_physical;
@@ -842,6 +836,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0(
                 mm_in1_sender_writer_args.push_back(last_out_block_w);
 
                 // padding args (WRITER)
+                mm_in1_sender_writer_args.push_back(last_out_num_blocks_w);
                 mm_in1_sender_writer_args.push_back(out_block_h / out_subblock_h);
                 mm_in1_sender_writer_args.push_back(out_subblock_h);
                 mm_in1_sender_writer_args.push_back(0);
@@ -855,6 +850,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0(
                 mm_in1_sender_writer_args.push_back(out_block_w);
 
                 // padding args (WRITER)
+                mm_in1_sender_writer_args.push_back(out_num_blocks_x);
                 mm_in1_sender_writer_args.push_back(out_block_h / out_subblock_h);
                 mm_in1_sender_writer_args.push_back(out_subblock_h);
                 mm_in1_sender_writer_args.push_back(0);
@@ -1492,19 +1488,11 @@ operation::ProgramWithCallbacks create_program_mcast_in1(
 
     // Parameters for last row, col, or block
     uint32_t last_per_core_M = M % per_core_M == 0 ? per_core_M : M % per_core_M;
-    uint32_t last_per_core_N = N % per_core_N == 0 ? per_core_N : N % per_core_N;
     uint32_t last_out_block_h = last_per_core_M % out_block_h == 0 ? out_block_h : last_per_core_M % out_block_h;
-    uint32_t last_out_block_w = last_per_core_N % out_block_w == 0 ? out_block_w : last_per_core_N % out_block_w;
+    uint32_t last_out_num_blocks_h = (last_per_core_M - 1) / out_block_h + 1;
     uint32_t last_block_num_nonzero_subblocks_h = (last_out_block_h - 1) / out_subblock_h + 1;
-    uint32_t last_block_num_nonzero_subblocks_w = (last_out_block_w - 1) / out_subblock_w + 1;
     uint32_t last_subblock_of_last_block_h =
         last_out_block_h % out_subblock_h == 0 ? out_subblock_h : last_out_block_h % out_subblock_h;
-    uint32_t last_subblock_of_last_block_w =
-        last_out_block_w % out_subblock_w == 0 ? out_subblock_w : last_out_block_w % out_subblock_w;
-    uint32_t last_block_padded_subblock_tiles_addr_skip =
-        output_single_tile_size * (out_subblock_w - last_subblock_of_last_block_w);
-    uint32_t last_block_padded_block_tiles_w_skip =
-        (out_subblock_w * out_subblock_h) * (out_block_w / out_subblock_w - last_block_num_nonzero_subblocks_w);
     uint32_t last_block_padded_block_tiles_h_skip =
         (out_block_h / out_subblock_h - last_block_num_nonzero_subblocks_h) * (out_block_w * out_subblock_h);
 
@@ -1541,6 +1529,7 @@ operation::ProgramWithCallbacks create_program_mcast_in1(
                 // padding args (READER)
                 (std::uint32_t)out_block_w,  // last_block_w
                 // padding args (WRITER)
+                (std::uint32_t)out_num_blocks_x,
                 (std::uint32_t)out_block_h / out_subblock_h,
                 (std::uint32_t)out_subblock_h,
                 (std::uint32_t)0,
@@ -1575,6 +1564,8 @@ operation::ProgramWithCallbacks create_program_mcast_in1(
 
             if (output_idx_y == num_blocks_y - 1) {
                 // padding args (WRITER)
+                mm_in1_receiver_writer_args.push_back(last_out_num_blocks_h);
+                mm_in1_receiver_writer_args.push_back(out_num_blocks_x);
                 mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                 mm_in1_receiver_writer_args.push_back(last_block_num_nonzero_subblocks_h);
                 mm_in1_receiver_writer_args.push_back(last_subblock_of_last_block_h);
@@ -1586,6 +1577,8 @@ operation::ProgramWithCallbacks create_program_mcast_in1(
                 mm_in1_receiver_writer_args.push_back(0);
             } else {
                 // padding args (WRITER)
+                mm_in1_receiver_writer_args.push_back(out_num_blocks_y);
+                mm_in1_receiver_writer_args.push_back(out_num_blocks_x);
                 mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                 mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                 mm_in1_receiver_writer_args.push_back(out_subblock_h);

--- a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_1d_program_factory.cpp
@@ -941,7 +941,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0(
                 writer_runtime_args[0] = src_buffer_b->address();
                 writer_runtime_args[6] = dst_buffer->address();
                 if (bias_tensor.has_value()) {
-                    writer_runtime_args[17] = (*bias_buffer)->address();
+                    writer_runtime_args[18] = (*bias_buffer)->address();
                 }
             }
 
@@ -1657,7 +1657,7 @@ operation::ProgramWithCallbacks create_program_mcast_in1(
                 sender_writer_runtime_args[0] = src_buffer_b->address();
                 sender_writer_runtime_args[6] = dst_buffer->address();
                 if (bias_tensor.has_value()) {
-                    sender_writer_runtime_args[17] = (*bias_buffer)->address();
+                    sender_writer_runtime_args[18] = (*bias_buffer)->address();
                 }
             }
 

--- a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp
@@ -845,6 +845,8 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
     uint32_t last_per_core_N = N % per_core_N == 0 ? per_core_N : N % per_core_N;
     uint32_t last_out_block_h = last_per_core_M % out_block_h == 0 ? out_block_h : last_per_core_M % out_block_h;
     uint32_t last_out_block_w = last_per_core_N % out_block_w == 0 ? out_block_w : last_per_core_N % out_block_w;
+    uint32_t last_out_num_blocks_h = (last_per_core_M - 1) / out_block_h + 1;
+    uint32_t last_out_num_blocks_w = (last_per_core_N - 1) / out_block_w + 1;
     uint32_t last_block_num_nonzero_subblocks_h = (last_out_block_h - 1) / out_subblock_h + 1;
     uint32_t last_block_num_nonzero_subblocks_w = (last_out_block_w - 1) / out_subblock_w + 1;
     uint32_t last_subblock_of_last_block_h =
@@ -1021,11 +1023,12 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                     (std::uint32_t)in1_idx * per_core_N + in0_idx * per_core_M * N  // out_tensor_start_tile_id
                 };
 
-                if (in1_idx == in1_end_idx) {
+                if (in1_idx == in1_end_idx) {  // right cores when no transpose_mcast
                     // padding args (READER)
                     mm_in1_sender_writer_args.push_back(last_out_block_w);
 
                     // padding args (WRITER)
+                    mm_in1_sender_writer_args.push_back(last_out_num_blocks_w);
                     mm_in1_sender_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_sender_writer_args.push_back(out_subblock_h);
                     mm_in1_sender_writer_args.push_back(0);
@@ -1039,6 +1042,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                     mm_in1_sender_writer_args.push_back(out_block_w);
 
                     // padding args (WRITER)
+                    mm_in1_sender_writer_args.push_back(out_num_blocks_x);
                     mm_in1_sender_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_sender_writer_args.push_back(out_subblock_h);
                     mm_in1_sender_writer_args.push_back(0);
@@ -1139,8 +1143,10 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                     (std::uint32_t)in1_idx * per_core_N + in0_idx * per_core_M * N  // out_tensor_start_tile_id
                 };
 
-                if (in1_idx == in1_end_idx and in0_idx == in0_end_idx) {
+                if (in1_idx == in1_end_idx and in0_idx == in0_end_idx) {  // bottom-right core when no transpose_mcast
                     // padding args (WRITER)
+                    mm_in1_receiver_writer_args.push_back(last_out_num_blocks_h);
+                    mm_in1_receiver_writer_args.push_back(last_out_num_blocks_w);
                     mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_receiver_writer_args.push_back(last_block_num_nonzero_subblocks_h);
                     mm_in1_receiver_writer_args.push_back(last_subblock_of_last_block_h);
@@ -1150,8 +1156,10 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                     mm_in1_receiver_writer_args.push_back(last_subblock_of_last_block_w);
                     mm_in1_receiver_writer_args.push_back(last_block_padded_subblock_tiles_addr_skip);
                     mm_in1_receiver_writer_args.push_back(last_block_padded_block_tiles_w_skip);
-                } else if (in0_idx == in0_end_idx) {
+                } else if (in0_idx == in0_end_idx) {  // bottom cores except bottom-right when no transpose_mcast
                     // padding args (WRITER)
+                    mm_in1_receiver_writer_args.push_back(last_out_num_blocks_h);
+                    mm_in1_receiver_writer_args.push_back(out_num_blocks_x);
                     mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_receiver_writer_args.push_back(last_block_num_nonzero_subblocks_h);
                     mm_in1_receiver_writer_args.push_back(last_subblock_of_last_block_h);
@@ -1161,8 +1169,10 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                     mm_in1_receiver_writer_args.push_back(out_subblock_w);
                     mm_in1_receiver_writer_args.push_back(0);
                     mm_in1_receiver_writer_args.push_back(0);
-                } else if (in1_idx == in1_end_idx) {
+                } else if (in1_idx == in1_end_idx) {  // right cores except bottom when no transpose_mcast
                     // padding args (WRITER)
+                    mm_in1_receiver_writer_args.push_back(out_num_blocks_y);
+                    mm_in1_receiver_writer_args.push_back(last_out_num_blocks_w);
                     mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_receiver_writer_args.push_back(out_subblock_h);
@@ -1174,6 +1184,8 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                     mm_in1_receiver_writer_args.push_back(last_block_padded_block_tiles_w_skip);
                 } else {
                     // padding args (WRITER)
+                    mm_in1_receiver_writer_args.push_back(out_num_blocks_y);
+                    mm_in1_receiver_writer_args.push_back(out_num_blocks_x);
                     mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_receiver_writer_args.push_back(out_block_h / out_subblock_h);
                     mm_in1_receiver_writer_args.push_back(out_subblock_h);

--- a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp
@@ -1063,6 +1063,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                 }
 
                 if (in1_is_sharded and in1_is_dram) {  // in1 is dram sharded
+                    uint32_t num_iter_index = mm_in1_sender_writer_args.size() + 1;
                     vc = vc == 3 ? 0 : vc + 1;
                     mm_in1_sender_writer_args.push_back(vc);
 
@@ -1121,7 +1122,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                             worker_core_stride = stride;
                         }
                     }
-                    mm_in1_sender_writer_args.insert(mm_in1_sender_writer_args.begin() + 20, num_iter);
+                    mm_in1_sender_writer_args.insert(mm_in1_sender_writer_args.begin() + num_iter_index, num_iter);
                 }
                 if (fuse_op) {
                     fused_op_signaler->push_matmul_fused_op_rt_args(mm_in1_sender_writer_args, true);
@@ -1268,7 +1269,7 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
                 writer_runtime_args[0] = src_buffer_b->address();
                 writer_runtime_args[6] = dst_buffer->address();
                 if (bias_tensor.has_value()) {
-                    writer_runtime_args[17] = (*bias_buffer)->address();
+                    writer_runtime_args[18] = (*bias_buffer)->address();
                 }
             }
 

